### ISSUE TITLE
Plate Elements and Excel Interface

### ISSUE THREAD

**Issue by DougAJ4** [JWock82/PyNite#68]
Hi, this is IDS from Eng-Tips

Could you let me know if you are happy if I link to this site from my blog, with a post summarising work described below?

Following the recent discussion, I have been experimenting with running your code from Excel, using pyxll, using code based on the   RectWall module, see sample output below.
![PyNite1-1](https://user-images.githubusercontent.com/7763499/95667929-877df900-0bb8-11eb-9b69-f4fc74252157.png)
![PyNite1-2](https://user-images.githubusercontent.com/7763499/95667938-9bc1f600-0bb8-11eb-941a-643875f04ee8.png)

I have compared midspan moments in both directions with results from Strand7, using their 4-node plate element (which does not include shear deformation).

The good news is that the moments are in near exact agreement (and deflections were also).

Not so good is the execution time, which is very slow  (see times in Column H).  The Strand7 completes in less than a second. I'm surprised the actual solve time is so slow, since you are using a Scipy solver, I suspect my time for solve may be including some pre or post processing.  Can you confirm if these results are similar to what you get?  If so it should be possible to greatly reduce them with some recoding of the bottlenecks.

Do you have any interface code for the Quad3D module?  Is this the one that was giving lower bending moments?  If so did you resolve the reason for that?

Please let me know if you would like me to upload any code or the spreadsheet, but the Python/Excel link does need pyxll, which is a commercial package.

**Comment by JWock82** [comment]
Go ahead and link to your blog post.

I am aware of the speed issue. My test model of 800 plates takes a few minutes to solve. It's something I plan to sort out eventually, but first I just want it all to work. The Quad3D elements are mostly finished. I need to add a few more features and clean up the code. Functionality first and then speed and performance. These are the MITC4 elements that were being discussed on the Eng-Tips forum.

The numpy solver is using the LAPACK routine which is supposedly very efficient. I don't think that's the issue. PyNite prints out messages at various steps and I noticed it seems to fly through that part. It seems to get hung up after it starts the stability check.

To interface with the Quad3D element use the `FEModel3D.AddQuad3D` method and the FEModel3D.AddQuadSurfacePressure` method.

**Comment by JWock82** [comment]
FYI: The rectangular plate elements seems to converge faster with a coarser mesh. If solve time is a problem you may want to use those for now. They match the Timoshenko solution pretty closely, which means they're not considering transverse shear deformations and may not be appropriate for thicker plates.

**Comment by JWock82** [comment]
..."Is this the one that was giving lower bending moments? If so did you resolve the reason for that?"

I think the plate formulation is correct, but I think the reason is 2-fold:

1. Transverse shear deformations increase the flexibility of the system, which leads to larger deflections and smaller loads.
2. The mesh may not have been fine enough. I went down to a 4"x4" mesh, and found that the stresses moved a little closer to the Timoshenko solution. I tried a 3"x3" mesh, but the solution speed was slow and my computer ran out of memory. Definitely some things I will eventually get around to addressing.

**Comment by DougAJ4** [comment]
I have now compared the Quad3D results with the Strand7 8node plate, which includes shear deflections:
![PyNite1-3](https://user-images.githubusercontent.com/7763499/95671367-f6237c80-0be1-11eb-9b60-e15bd3f1a73a.png)
![PyNite1-4](https://user-images.githubusercontent.com/7763499/95671371-fc195d80-0be1-11eb-9993-189f2b18cac8.png)

The match is not as good as the 4 node elements, but that's to be expected with a different element type.  All the results except the edge nodes are within about 2%, which I'd say is as good as would be expected with different element types.  The edge moments are significantly lower, which suggests to me that there is an error extracting the stresses for these results.  Since the deflections are a good match, it must be in the post-processing, rather than the solution of the main matrix.

**Comment by JWock82** [comment]
After reading studying it closely for the past week and comparing the results to known solutions for the MITC4 element, I'm almost certain the element is formulated correctly, and that this is a limitation with the MITC4 element itself at boundaries. Other popular programs that use the MITC4 element appear to have this issue as well. See [example B-04 here](https://enercalc.com/ec3d_verification/enercalc3dverifications.html) for one example.

We're treading dangerous water here by comparing results to other programs. I don't want to get accused of reverse-engineering anything. Maybe I'm over-thinking it because based on that link I posted it looks like that's how these other programs verify themselves. I try to verify PyNite with published solutions instead, which is why I've been using the Timoshenko solutions. I also have PCA published solutions and BLM monograph No. 27 that I'd like to compare to. I don't know that any of these include transverse shear deformations though which has made this trickier.

**Comment by JWock82** [comment]
All that being said, the MITC4 element does perform very well where others do not. It handles being distorted much better that other elements. That's the reason I selected this element to begin with.

**Comment by JWock82** [comment]
Quad elements were recalculating the local coordinate system multiple times. I just corrected this. More speed improvements to come. A big thank you to @tamalone1 who clued me in to how to profile this program. I know exactly where the slow-downs are and have some ideas about how to go about fixing them. I plan to start incorporating the fixes this next week.

**Comment by JWock82** [comment]
The good news: It's solving much much faster now.
The bad news: In my zeal to increase speed I seem to have broken the quad element's stiffness matrix. Looks like I'll be sorting through this now.

**Comment by JWock82** [comment]
Fixed it!

**Comment by JWock82** [comment]
@DougAJ4 How did you come up with the stresses at each node point? Did you average the results from each of the plate corners at the node? Your MX value at X=6  of -1.478252581 appears to be an average of the results for all plate corners adjoining the node.

> 
> 
> I have now compared the Quad3D results with the Strand7 8node plate, which includes shear deflections:
> ![PyNite1-3](https://user-images.githubusercontent.com/7763499/95671367-f6237c80-0be1-11eb-9b60-e15bd3f1a73a.png)
> ![PyNite1-4](https://user-images.githubusercontent.com/7763499/95671371-fc195d80-0be1-11eb-9993-189f2b18cac8.png)
> 
> The match is not as good as the 4 node elements, but that's to be expected with a different element type. All the results except the edge nodes are within about 2%, which I'd say is as good as would be expected with different element types. The edge moments are significantly lower, which suggests to me that there is an error extracting the stresses for these results. Since the deflections are a good match, it must be in the post-processing, rather than the solution of the main matrix.

**Comment by DougAJ4** [comment]
Hi Craig,
The listed moments are generated with:
    ...
    myWall = RectWall2.RectWall(width, height, t, E, nu, meshsize,
*Supports)
    # Get loads from the spreadsheet
    loady0, loady1 = Loads[0,:]
    load0, load1 = Loads[1,:]
    myWall.add_load(loady0, loady1, load0, load1)
    myWall.analyze()
    ...
    rtn = myWall.plot_forces('Mx')

So the forces are generated by your code, but I understand that they do
average the moment at each node.

I am (very slowly) working through your code to understand how it all
works.  I'm still not convinced that the edge moments are extracted
correctly, but I will report back one way or the other whan I have finished
working on it.

Doug

**Comment by JWock82** [comment]
The “plot_forces” method plots smoothed contours (averaged at the nodes). I got the visualization module plotting unsmoothed stresses on the 3D view. For the Plate3D element the stresses looked really good, but when I ran the same problem for the Quad3D the stresses were nearly constant over each plate. I think it’s reporting plate center stresses at the corners. If you average the stresses between adjacent plates, you get the correct answers at the shared corners. At the edges there are no adjacent plates to average with, so it essentially reports the plate center stresses at the supports. What’s weird is the unsmoothed deflected shape color map looks great. I’ve tried 2 different methods for extracting corner stresses, and I get the same results. I’m thinking the MITC4 is bad at reporting corner stresses. I will continue looking into it.

**Comment by DougAJ4** [comment]
...but when I ran the same problem for the Quad3D the stresses were nearly constant over each plate. I think it’s reporting plate center stresses at the corners.

That's consistent with what I found.  I'm wondering if there is an inconsistency in the interpolation procedure between the global stiffness matrix, and extraction of the local forces for each quad.

**Comment by JWock82** [comment]
I'm going to close this issue out. I've got the program's speed moving much faster, and I plan to continue isolating slow points in the program as I continue developing. Rectangular plate speeds need attention next.

Regarding the stress results. I've updated the rectangular wall panel example for quads with a full discussion of what I believe is going on. After reading some more, and finding articles on extrapolating stresses I believe the elements are working correctly and performing very well. See the example for a full discussion of what's going on.
